lib/Modem.ts

```diff
@@ -0,0 +1,829 @@
+import * as pdu from 'node-pdu';
+import { SerialPort } from 'serialport';
+import {
+	CheckModemSuccess,
+	CommandResponse,
+	ModemConstructorOptions,
+	ModemMode,
+	PduSms,
+	SendSMSFailed,
+	SendSMSSuccess,
+	SerialPortOptions,
+	TxtSms
+} from './types';
+import { Command } from './utils/Command';
+import { Events } from './utils/Events';
+import * as prTypes from './utils/internalTypes';
+import { ModemOptions } from './utils/internalTypes';
+
+export class Modem {
+	// options
+	private _mode: ModemMode;
+	private readonly pinCode: string | null;
+	private readonly options: ModemOptions;
+
+	// system
+	private readonly port: SerialPort;
+	readonly events = new Events();
+
+	// command queue
+	private prioQueue: (Command | prTypes.CmdSequence)[] = [];
+	private queue: (Command | prTypes.CmdSequence)[] = [];
+	private _receivedData = '';
+	private _receivedCmdResponse: CommandResponse = [];
+	private _isLocked = false;
+	private _interval: NodeJS.Timer | null = null;
+
+	constructor(targetDevice: string, mode: ModemMode = 'PDU', options: ModemConstructorOptions = {}) {
+		this._mode = mode;
+		this.pinCode = options.pin || null;
+
+		this.options = {
+			autoDeleteOnReceive: options.autoDeleteOnReceive !== undefined ? options.autoDeleteOnReceive : false,
+			enableConcatenation: options.enableConcatenation !== undefined ? options.enableConcatenation : false,
+			customInitCommand: options.customInitCommand !== undefined ? options.customInitCommand : null,
+			autoInitOnOpen: options.autoInitOnOpen !== undefined ? options.autoInitOnOpen : true,
+			cnmiCommand: options.cnmiCommand !== undefined ? options.cnmiCommand : 'AT+CNMI=2,1,0,2,1'
+		};
+
+		const serialPortOptions: SerialPortOptions = Object.assign(
+			// defaults
+			{
+				baudRate: 9600,
+				dataBits: 8,
+				stopBits: 1,
+				highWaterMark: 16384,
+				parity: 'none',
+				rtscts: false,
+				xon: false,
+				xoff: false
+			},
+			// options
+			options.serialPortOptions,
+			// overide options
+			{
+				path: targetDevice,
+				autoOpen: false
+			}
+		);
+
+		this.port = new SerialPort(serialPortOptions);
+		this.port.on('data', (data) => this._dataReceived(data));
+	}
+
+	/*
+	 * getter
+	 */
+
+	get targetDevice() {
+		return this.port.path;
+	}
+
+	get isOpen() {
+		return this.port.isOpen;
+	}
+
+	get mode() {
+		return this._mode;
+	}
+
+	get queueLength() {
+		return this.prioQueue.length + this.queue.length;
+	}
+
+	/*
+	 * command queue
+	 */
+
+	private pushToQue(cmd: Command | prTypes.CmdSequence, prio = false) {
```

Comment:
Typo? pushToQueue?

Reply:
I wanted to save a few characters 😅. But actually you can write it out that it's right

Reply:
Is hopefully improved in the next update